Partial EveryVoice integration with WW

roedoejet opened the pull request:
This PR does the following:

- Create an EveryVoice module project within the wordweaver mono-repo
- Creates an EveryVoice component and directive for adding UI components to an Angular project
- Creates an EveryVoice service to play audio (this is just a stub that uses the system default TTS while @dlothian works on the real thing)
- Creates a configuration token that can be used to inject a configuration into the EveryVoice module when importing into a new project. The configuration has been added to the environment.ts to allow users to configure from their environment file.
- Adds UI integration in WW using the EveryVoice components and service
- UI integration included in wordmaker, and all tableviewer views
- The environment.ts file allows the TTS to be disabled entirely
- Adds [documentation](https://github.com/roedoejet/wordweaver/blob/dev.ap/tts-library/projects/every-voice/README.md) on how to use it

Next steps:

- [ ] Add a service that is able to connect to an EveryVoice gradio instance
- [ ] Publish the EveryVoice module to npm so that we can easily add TTS to any other Angular projects (question: any thoughts on the publishing namespace for this?)

roedoejet committed:
feat: initial build of everyvoice library

roedoejet committed:
feat: add component and allow configurable service

roedoejet committed:
feat: connect wordweaver with everyvoice plugin

roedoejet committed:
fix(tv): fix tableviewer tree view pronouns

roedoejet committed:
feat: add tts to shared module and tier

roedoejet committed:
feat(tts): add tts to all tableviewer views

roedoejet committed:
feat: allow tts to be enabled and disabled when importing the module

roedoejet committed:
refactor: allow ev tts config in environment file

roedoejet committed:
fix(tests): fix spec tests

roedoejet committed:
fix(tests): fix every-voice test suite

roedoejet committed:
fix(ci): update build scripts to build ev first

roedoejet committed:
docs: add everyvoice installation documentation

roedoejet committed:
fix(tv): add audio ui for single verb grid view

deltork commented on projects/every-voice/src/lib/every-voice.service.ts:
```diff
@@ -0,0 +1,27 @@
+    @Optional() @Inject(EVERY_VOICE_CONFIG) config: EveryVoiceConfig
+  ) {
+    this.apiUrl = config?.apiUrl ?? "https://default.api/tts";
+    this.enableTTS = config?.enableTTS ?? true;
```
It would be better if the default is **true** only if you have defined an **apiUrl** in your **config**.

```suggestion
    this.enableTTS = (config?.enableTTS !== undefined ? config.enableTTS : (config?.apiUrl?.length > 0));
```

roedoejet replied:
great point - thanks @deltork

deltork commented on projects/every-voice/src/lib/every-voice/every-voice.component.html:
```diff
@@ -0,0 +1,3 @@
+<button mat-icon-button (click)="onClick()">
+  <mat-icon>{{ icon }}</mat-icon>
```
Should this button's visual state change when synthesis is progressing or playing? This is to prevent multiple clicks by the user because it will appear to be doing nothing.  Use the button to stop play
```suggestion
  <mat-icon>{{ isPlaying? stopIcon : icon }}</mat-icon>
```

roedoejet replied:
good idea. done.

deltork commented on projects/every-voice/src/lib/every-voice/every-voice.component.ts:
```diff
@@ -0,0 +1,25 @@
+export class EveryVoiceComponent implements OnInit {
+  @Input() textToGenerate: string;
+  @Input() icon = "volume_up";
+
```
Should this component be visible if EveryVoiceService is disabled ?
```suggestion
@Input() stopIcon = "stop";
isPlaying = false;
```

roedoejet replied:
fair enough - I did this in the wordweaver code, but may as well actually just add `[hidden]` on the component itself

deltork commented on projects/every-voice/src/lib/every-voice/every-voice.component.ts:
```diff
@@ -0,0 +1,25 @@
+
+  onClick() {
+    if (this.textToGenerate) {
+      this.tts.playSound(this.textToGenerate);
```
```suggestion
      if (this.isPlaying){
        this.tts.stop();
        this.isPlaying = false;
      }else{
        this.isPlaying = true
        this.tts.playSound(this.textToGenerate).error((error)=>{/*do something with the error like toast notification*/ console.error(error)}).finally(()=>this.isPlaying = false);
      }
      
```

roedoejet replied:
I assume .catch - but yes this is a good suggestion. I'm not going to handle the error though, since this is really just a placeholder service.

deltork commented on projects/every-voice/src/lib/every-voice.service.ts:
```diff
@@ -0,0 +1,27 @@
+    this.apiUrl = config?.apiUrl ?? "https://default.api/tts";
+    this.enableTTS = config?.enableTTS ?? true;
+  }
+  playSound(text: string) {
```
This function needs to return a promise...
```suggestion
  playSound(text: string) :Promise<void>{
    return new Promise((resolve, reject) => {
      if (this.enableTTS) {
        console.log(`Calling ${this.apiUrl} with text:`, text);
        const utterance = new SpeechSynthesisUtterance(text);
        utterance.onend = () => resolve();
        utterance.onerror = (event) => reject(event.error);
        window.speechSynthesis.speak(utterance);
      } else {
        reject("Text-to-Speech is disabled");
      }
    });
    }
```

deltork reviewed (changes requested):
Looks good!

dlothian reviewed (approved):
I focused on reviewing what was related to UI integration using the ev components and service, it all looks good and intuitive to use. Looking forward to building on it. Good job!

BY SUGGESTION OF AKWIRATÉKHA':
Change it so that in addition to the audio icon, there is an invisible box around the icon and the text that is clickable for audio. 

-- edit -- 
forgive me forgot to give my suggestions for the npm module name for publishing:

    Angul-eryVoice Module (Ang-eryVoice?)
    EV-ular Module
    Angular-to-speech (EV)

roedoejet committed:
fix: make Del's suggested changes

returns a promise, changes icon when playing or not, change default for enableTTS

roedoejet committed:
refactor: change package name to include organization name

roedoejet committed:
feat: allow directives to pull text content

add directives to wordweaver
